File: read-config.js
#!/usr/bin/env node
const fs = require("fs");
const path = require("path");
const http = require("http");
const https = require("https");
const { URL } = require("url");

function fileExists(p) {
  try {
    return fs.existsSync(p);
  } catch (e) {
    return false;
  }
}

function loadConfig() {
  const cfg = {};
  if (process.env.CENTRAL_MCP_SERVER_URL)
    cfg.centralMcpServerUrl = process.env.CENTRAL_MCP_SERVER_URL;
  if (process.env.CENTRAL_MCP_SERVER_TOKEN)
    cfg.centralMcpServerToken = process.env.CENTRAL_MCP_SERVER_TOKEN;
  if (process.env.CENTRAL_MCP_JWT_SECRET)
    cfg.centralMcpJwtSecret = process.env.CENTRAL_MCP_JWT_SECRET;

  const candidates = [
    "C:/central-mcp-config.json",
    path.join(process.cwd(), "central-mcp-config.json"),
  ];
  for (const c of candidates) {
    if (fileExists(c)) {
      try {
        const raw = fs.readFileSync(c, "utf8");
        const parsed = JSON.parse(raw);
        // merge without overwriting env values
        if (!cfg.centralMcpServerUrl && parsed.centralMcpServerUrl)
          cfg.centralMcpServerUrl = parsed.centralMcpServerUrl;
        if (!cfg.centralMcpServerToken && parsed.centralMcpServerToken)
          cfg.centralMcpServerToken = parsed.centralMcpServerToken;
        if (!cfg.centralMcpJwtSecret && parsed.centralMcpJwtSecret)
          cfg.centralMcpJwtSecret = parsed.centralMcpJwtSecret;
        if (!cfg.secrets && parsed.secrets) cfg.secrets = parsed.secrets;
      } catch (e) {
        console.error("failed to read/parse", c, e && e.message);
      }
      break;
    }
  }
  return cfg;
}

function httpRequest(urlStr, opts = {}) {
  return new Promise((resolve, reject) => {
    const u = new URL(urlStr);
    const lib = u.protocol === "https:" ? https : http;
    const req = lib.request(
      {
        hostname: u.hostname,
        port: u.port,
        path: u.pathname + (u.search || ""),
        method: opts.method || "GET",
        headers: opts.headers || {},
        timeout: opts.timeout || 5000,
      },
      (res) => {
        let data = "";
        res.on("data", (d) => (data += d.toString()));
        res.on("end", () => {
          res.body = data;
          resolve(res);
        });
      }
    );
    req.on("error", reject);
    if (opts.body) req.write(opts.body);
    req.end();
  });
}

async function requestJwt(serverUrl, serverToken) {
  if (!serverUrl) throw new Error("serverUrl empty");
  if (!serverToken) throw new Error("serverToken empty");
  const url = serverUrl.replace(/\/+$/, "") + "/token";
  const res = await httpRequest(url, {
    method: "POST",
    headers: { Authorization: "Bearer " + serverToken },
  });
  if (res.statusCode !== 200)
    throw new Error(`token request failed ${res.statusCode}: ${res.body}`);
  const parsed = JSON.parse(res.body);
  if (!parsed.access_token) throw new Error("no access_token in response");
  return parsed.access_token;
}

async function getSecret(serverUrl, jwt, name) {
  const url =
    serverUrl.replace(/\/+$/, "") + "/secrets/" + encodeURIComponent(name);
  const res = await httpRequest(url, {
    method: "GET",
    headers: { Authorization: "Bearer " + jwt },
  });
  if (res.statusCode !== 200)
    throw new Error(`secret request failed ${res.statusCode}: ${res.body}`);
  const parsed = JSON.parse(res.body);
  return parsed.value;
}

(async function main() {
  const cfg = loadConfig();
  const secretName = process.argv[2] || "mySecretKey";

  if (!cfg.centralMcpServerUrl) {
    console.error(
      "no server URL configured. set CENTRAL_MCP_SERVER_URL or place central-mcp-config.json"
    );
    process.exit(2);
  }

  if (!cfg.centralMcpServerToken) {
    console.error(
      "no server token configured. set CENTRAL_MCP_SERVER_TOKEN or place central-mcp-config.json"
    );
    process.exit(2);
  }

  try {
    const jwt = await requestJwt(
      cfg.centralMcpServerUrl,
      cfg.centralMcpServerToken
    );
    // console.log('jwt', jwt);
    const val = await getSecret(cfg.centralMcpServerUrl, jwt, secretName);
    console.log(val);
  } catch (e) {
    console.error("failed:", e && e.message);
    process.exit(1);
  }
})();

File: read_config.go
package main

import (
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"
)

type Config struct {
	CentralMcpServerUrl   string            `json:"centralMcpServerUrl"`
	CentralMcpServerToken string            `json:"centralMcpServerToken"`
	CentralMcpJwtSecret   string            `json:"centralMcpJwtSecret"`
	Secrets               map[string]string `json:"secrets"`
}

func fileExists(p string) bool {
	_, err := os.Stat(p)
	return err == nil
}

func loadConfig() (*Config, error) {
	cfg := &Config{}
	// First, read environment overrides (they take precedence)
	if v := os.Getenv("CENTRAL_MCP_SERVER_URL"); v != "" {
		cfg.CentralMcpServerUrl = v
	}
	if v := os.Getenv("CENTRAL_MCP_SERVER_TOKEN"); v != "" {
		cfg.CentralMcpServerToken = v
	}
	if v := os.Getenv("CENTRAL_MCP_JWT_SECRET"); v != "" {
		cfg.CentralMcpJwtSecret = v
	}

	// Candidate file locations (prefer C:\ if present to match server behavior)
	candidates := []string{
		`C:\central-mcp-config.json`,
		"central-mcp-config.json",
	}

	for _, p := range candidates {
		if fileExists(p) {
			b, err := os.ReadFile(p)
			if err != nil {
				return nil, fmt.Errorf("failed to read %s: %w", p, err)
			}
			var fcfg Config
			if err := json.Unmarshal(b, &fcfg); err != nil {
				return nil, fmt.Errorf("failed to parse %s: %w", p, err)
			}
			// Merge: environment values already set take precedence
			if cfg.CentralMcpServerUrl == "" {
				cfg.CentralMcpServerUrl = fcfg.CentralMcpServerUrl
			}
			if cfg.CentralMcpServerToken == "" {
				cfg.CentralMcpServerToken = fcfg.CentralMcpServerToken
			}
			if cfg.CentralMcpJwtSecret == "" {
				cfg.CentralMcpJwtSecret = fcfg.CentralMcpJwtSecret
			}
			if cfg.Secrets == nil {
				cfg.Secrets = fcfg.Secrets
			}
			return cfg, nil
		}
	}

	// No file found, return whatever we have from env (may be empty)
	return cfg, nil
}

func requestJWT(serverURL, serverToken string) (string, error) {
	if serverURL == "" {
		return "", errors.New("server URL is empty")
	}
	if serverToken == "" {
		return "", errors.New("server token is empty")
	}
	client := &http.Client{Timeout: 5 * time.Second}
	req, err := http.NewRequest("POST", strings.TrimRight(serverURL, "/")+"/token", nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("Authorization", "Bearer "+serverToken)
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		b, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("token request failed %d: %s", resp.StatusCode, string(b))
	}
	var body struct {
		AccessToken string `json:"access_token"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
		return "", err
	}
	if body.AccessToken == "" {
		return "", errors.New("no access_token in response")
	}
	return body.AccessToken, nil
}

func getSecret(serverURL, jwt, name string) (string, error) {
	if serverURL == "" {
		return "", errors.New("server URL is empty")
	}
	client := &http.Client{Timeout: 5 * time.Second}
	req, err := http.NewRequest("GET", strings.TrimRight(serverURL, "/")+"/secrets/"+urlEscape(name), nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("Authorization", "Bearer "+jwt)
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	b, _ := io.ReadAll(resp.Body)
	if resp.StatusCode != 200 {
		return "", fmt.Errorf("secret request failed %d: %s", resp.StatusCode, string(b))
	}
	var out struct {
		Name  string `json:"name"`
		Value string `json:"value"`
	}
	if err := json.Unmarshal(b, &out); err != nil {
		return "", err
	}
	return out.Value, nil
}

func urlEscape(s string) string {
	// simple escape for path segment
	return strings.ReplaceAll(s, " ", "%20")
}

func mask(s string) string {
	if s == "" {
		return ""
	}
	if len(s) <= 4 {
		return "****"
	}
	return s[:2] + strings.Repeat("*", len(s)-4) + s[len(s)-2:]
}

func main() {
	secretFlag := flag.String("secret", "", "Secret name to fetch from central server")
	showCfg := flag.Bool("show", false, "Print resolved configuration (masked)")
	flag.Parse()

	cfg, err := loadConfig()
	if err != nil {
		fmt.Fprintln(os.Stderr, "failed to load config:", err)
		os.Exit(1)
	}

	if *showCfg {
		fmt.Println("resolved config:")
		fmt.Println("  serverUrl:", cfg.CentralMcpServerUrl)
		fmt.Println("  serverToken:", mask(cfg.CentralMcpServerToken))
		fmt.Println("  jwtSecret:", mask(cfg.CentralMcpJwtSecret))
		os.Exit(0)
	}

	if *secretFlag == "" {
		// if no secret requested print available local secrets from file
		if cfg.Secrets != nil && len(cfg.Secrets) > 0 {
			fmt.Println("local secrets:")
			for k := range cfg.Secrets {
				fmt.Println(" -", k)
			}
			os.Exit(0)
		}
		fmt.Println("no secret requested and no local secrets available; use -secret NAME to fetch from server")
		os.Exit(0)
	}

	// Need server URL and server token (either from env/file)
	if cfg.CentralMcpServerUrl == "" {
		fmt.Fprintln(os.Stderr, "no server URL configured (env CENTRAL_MCP_SERVER_URL or central-mcp-config.json)")
		os.Exit(2)
	}
	if cfg.CentralMcpServerToken == "" {
		fmt.Fprintln(os.Stderr, "no server token configured (env CENTRAL_MCP_SERVER_TOKEN or central-mcp-config.json)")
		os.Exit(2)
	}

	jwt, err := requestJWT(cfg.CentralMcpServerUrl, cfg.CentralMcpServerToken)
	if err != nil {
		fmt.Fprintln(os.Stderr, "failed to obtain JWT:", err)
		os.Exit(3)
	}

	val, err := getSecret(cfg.CentralMcpServerUrl, jwt, *secretFlag)
	if err != nil {
		fmt.Fprintln(os.Stderr, "failed to fetch secret:", err)
		os.Exit(4)
	}
	fmt.Printf("%s\n", val)
}
